hack/update-knative-components.go

```diff
@@ -0,0 +1,258 @@
+	fs := bufio.NewScanner(f)
+	fs.Split(bufio.ScanLines)
+	for fs.Scan() {
+		// trim white space -> split the line via '='
+		lineArr := strings.Split(strings.TrimSpace(fs.Text()), "=")
+		if len(lineArr) != 2 {
+			continue
+		}
+		// add "=" for consise matching of the value assignment -- in case the value
+		// is used elsewhere in the file for any reason.
+		prefix := lineArr[0] + "="
+		// trim '"' and 'v' from the value for robustness (will be added back later)
+		// "v1.2.3" or v1.2.3 or "1.2.3" or v1.2.3 --> 1.2.3
+		val := strings.TrimFunc(lineArr[1], func(r rune) bool {
+			return r == '"' || r == 'v'
+		})
+
+		switch prefix {
+		case knSrvPrefix:
+			srv = "v" + val
+		case knEvtPrefix:
+			evt = "v" + val
+		case knCtrPrefix:
+			ctr = "v" + val
+		}
+
+		// if all values are acquired, no need to continue
+		if srv != "" && evt != "" && ctr != "" {
+			break
+		}
+	}
+	return
+}
+
+// try updating the version of component named by "repo" via 'sed'
+func tryUpdateFile(prefix, newV, oldV string) (bool, error) {
+	fmt.Printf("> try updating %s(%s -> %s)...", prefix, oldV, newV)
+	quoteWrap := func(s string) string {
+		if !strings.HasPrefix(s, "\"") {
+			return "\"" + s + "\""
+		}
+		return s
+	}
+	if newV != oldV {
+		fmt.Println("updating")
+		cmd := exec.Command("sed", "-i", "-e", "s/"+prefix+quoteWrap(oldV)+"/"+prefix+quoteWrap(newV)+"/g", file)
+		err := cmd.Run()
+		if err != nil {
+			return false, fmt.Errorf("error while updating file: %s", err)
+		}
+		return true, nil
+	}
+	return false, nil
+}
+
+// prepare branch for PR via git commands
+func prepareBranch(branchName string) error {
+	fmt.Println("> preparing branch")
+	cmd := exec.Command("bash", "-c", fmt.Sprintf(`
+		git config --local user.email "automation@knative.team" &&
+		git config --local user.name "Knative Automation" &&
+		git switch -c %s &&
+		git add %s &&
+		git commit -m "update components" &&
+		git push origin %s
+	`, branchName, file, branchName))
+	return cmd.Run()
+}
+
+// create a PR for the new updates
+func createPR(ctx context.Context, client *github.Client, title string, branchName string) error {
+	fmt.Print("> creating PR...")
+	body := fmt.Sprintf("%s\n/assign @gauron99", title)
+
+	newPR := github.NewPullRequest{
+		Title:               github.Ptr(title),
+		Base:                github.Ptr("main"),
+		Head:                github.Ptr(branchName),
+		Body:                github.Ptr(body),
+		MaintainerCanModify: github.Ptr(true),
+	}
+	pr, _, err := client.PullRequests.Create(ctx, "knative", "func", &newPR)
+	if err != nil {
+		fmt.Printf("PR looks like this:\n%#v\n", pr)
+		fmt.Printf("err: %s\n", err)
+		return err
+	}
+	fmt.Println("ready")
+	return nil
+}
+
+// returns true when PR with given title already exists in knative/func repo
+// otherwise false. Returns an error if occurred, otherwise nil.
+func prExists(ctx context.Context, c *github.Client, title string) (bool, error) {
+	opt := &github.PullRequestListOptions{State: "open"}
```

Comment:
Technically you should iterate over multiple "pages", but chances are the PR is in the first "page" (first 30 entries).

Reply:
exactly I dont think we are going to run into a 'second page PR' with this since we will be merging it 90% of the time without fail and the rest with some version bumps probably